pala.js

```diff
@@ -126,96 +191,181 @@ rules.sendRequestVote = (model, server, peer) => {
+      server.epoch += 1
+      server.epochChangingMessage.isCanBeSent = true
+      server.state = server.id === getNewProposerIdxFromServer(server.epoch) ? SERVER_STATES.leader : SERVER_STATES.follower
+      clearServer(model, server)
+      return
+    }
+    if(!server.epochChangingMessage.isCanBeSent) {
+      server.epochChangingMessage.isCanBeSent = false
+      server.epoch = server.epochChangingMessage.serverEpochToChange
+    }
+    server.state = server.id === getNewProposerIdxFromServer(server.epoch) ? SERVER_STATES.leader : SERVER_STATES.follower
     clearServer(model, server)
-    server.term += 1
   }
 };
 
 const clearServer = (model, server) => {
+  server.isRecoveryMessageSent = false
   server.votedFor = null
-  server.electionAlarm =
-      server.state === SERVER_STATES.leader ||
-      server.state === SERVER_STATES.stopped
-          ? 0
-          : makeElectionAlarm(model.time)
-  server.voteGranted = util.makeMap(server.peers, false)
+  server.electionAlarm = server.state === SERVER_STATES.stopped ? 0 : makeElectionAlarm(model.time)
   server.rpcDue = util.makeMap(server.peers, model.time + RPC_TIMEOUT)
-  server.heartbeatDue = util.makeMap(server.peers, model.time + ELECTION_TIMEOUT / 2)
+  server.heartbeatDue = util.makeMap(server.peers, 0)
 }
 
-rules.sendAppendEntries = (model, server, peer) => {
-  if (server.state === SERVER_STATES.leader &&
-      (server.heartbeatDue[peer] <= model.time ||
-       (server.nextIndex[peer] <= server.log.length &&
-        server.rpcDue[peer] <= model.time))) {
-    const prevIndex = server.nextIndex[peer] - 1;
-    let lastIndex = Math.min(prevIndex + BATCH_SIZE,
-                             server.log.length);
-    if (server.matchIndex[peer] + 1 < server.nextIndex[peer])
-      lastIndex = prevIndex;
+rules.sendAppendEntries = (model, server) => {
+  server.peers.forEach(peer => {
+    if(server.state === SERVER_STATES.leader) {
+      if ((server.heartbeatDue[peer] <= model.time ||
+          (server.nextIndex[peer] <= server.log.length &&
+              server.rpcDue[peer] <= model.time))) {
+        const prevIndex = server.nextIndex[peer] - 1;
+        let lastIndex = Math.min(prevIndex + BATCH_SIZE,
+            server.log.length);
+        if (server.matchIndex[peer] + 1 < server.nextIndex[peer])
+          lastIndex = prevIndex;
+        server.votesForBlock = util.makeMap(server.peers, -1)
+        sendRequest(model, {
+          from: server.id,
+          to: peer,
+          type: REQUEST_TYPES.appendEntries,
+          epoch: server.epoch,
+          prevIndex: prevIndex,
+          blockToVote: server.blockToVote,
+          prevTerm: logTerm(server.log, prevIndex),
+          entries: server.log.slice(prevIndex, lastIndex),
+          commitIndex: Math.min(server.commitIndex, lastIndex)
+        });
+        server.rpcDue[peer] = model.time + RPC_TIMEOUT;
+        server.heartbeatDue[peer] = model.time + ELECTION_TIMEOUT / 2;
+      }
+    }
+  })
+};
```

Comment:
hard to read code sections
I suggest 
1.Format the code(add spaces)
2. Make early exit conditions
``` if (server.state !== SERVER_STATES.leader) {
    return
  }```
  before foreach, so as not to check it at each iteration